language_service/src/compilation.rs

```diff
@@ -0,0 +1,258 @@
+        for (name, contents) in cells {
+            trace!("compiling cell {name}");
+            let increment = compiler
+                .compile_fragments(name, contents, |cell_errors| {
+                    errors.extend(cell_errors);
+                    Ok(()) // accumulate errors without failing
+                })
+                .expect("compile_fragments_acc_errors should not fail");
+
+            compiler.update(increment);
+        }
+
+        let (package_store, package_id) = compiler.into_package_store();
+
+        Self {
+            package_store,
+            user: package_id,
+            errors,
+            kind: CompilationKind::Notebook,
+        }
+    }
+
+    /// Gets the `CompileUnit` associated with user (non-library) code.
+    pub fn user_unit(&self) -> &CompileUnit {
+        self.package_store
+            .get(self.user)
+            .expect("expected to find user package")
+    }
+
+    /// Returns the package and `SourceMap` offset associated with a code location.
+    pub(crate) fn resolve_offset(&self, source_name: &str, offset: u32) -> (&AstPackage, u32) {
+        let unit = self.user_unit();
+
+        // Map the file offset into a SourceMap offset
+        let offset = unit
+            .sources
+            .find_by_name(source_name)
+            .expect("source should exist in the source map")
+            .offset
+            + offset;
+
+        let ast_package = &unit.ast;
+        (ast_package, offset)
+    }
+
+    /// Regenerates the compilation with the same sources but the passed in configuration options.
+    pub fn recompile(&mut self, package_type: PackageType, target_profile: TargetProfile) {
+        let sources = self.user_source_contents();
+
+        let new = match self.kind {
+            CompilationKind::OpenDocument => {
+                assert!(sources.len() == 1);
+                Self::new_open_document(sources[0].0, sources[0].1, package_type, target_profile)
+            }
+            CompilationKind::Notebook => Self::new_notebook(sources.into_iter()),
+        };
+        self.package_store = new.package_store;
+        self.user = new.user;
+        self.errors = new.errors;
+    }
+
+    /// Returns the original sources that were used to create the compilation.
+    fn user_source_contents(&self) -> Vec<(&str, &str)> {
+        let sources = &self.user_unit().sources;
+
+        successors(sources.find_by_offset(0), |last| {
+            sources
+                .find_by_offset(
+                    u32::try_from(last.contents.len()).expect("source contents should fit in u32")
+                        + 1,
+                )
+                .and_then(|s| {
+                    if s.offset == last.offset {
+                        None
+                    } else {
+                        Some(s)
+                    }
+                })
+        })
+        .map(|s| (s.name.as_ref(), s.contents.as_ref()))
+        .collect()
+    }
+}
+
+pub(crate) trait Lookup {
```

Comment:
I didn't see where this trait is used beyond `Compilation`. Is it used somewhere else?

Reply:
No, it's not used anywhere else. It's only hear to provide a bit of code organization - I feel that all these methods belong together and this felt like a nice way to group them together. 

I expect `Compilation` and its impl to grow and grow in the future, and I'm slightly concerned about it veering into [God object](https://en.wikipedia.org/wiki/God_object) territory. I wanted to take a step towards breaking it up without necessarily doing all the work to separate out the data yet. I might perhaps pull out `package_store` and `user_package` into their own struct definition one day, and on that day `Lookup` can just become an impl for _that_ struct.